Xenomorph embryos now grow based on consistent timers instead of 100% RNG (3% chance every 2s)

KathrinBailey opened the pull request:
## About The Pull Request
• Adds documentation to alien_embryo.dm
• Refactors how alien embryos increase their stage var, aka their cute ickle baby development stage. Now it increases stage every minute, which means 5 minutes to grow as opposed to taking 15 seconds to the end of time to grow.
• Growth time is now defined by a variable, so admins can var-edit it in game and future coders can easily change the variable to make it take more or less time to grow. It also means it's a simple one-line edit if any of TG's plentiful downstreams want to change this for their server.

## Why It's Good For The Game
This adds consistency to Xenomorphs and removes the frustration of your entire playthrough as an antagonist potentially being down to literal random chance to when you get a new larva buddy. Will it take two minutes, or ten? You have no idea. But your nest is about to be broken into and this larva could come out in the next second or in a few more minutes.

I tested how long it takes to grow larva using the old system, and the average was roughly around five minutes for a larva to hatch, hence why I chose a static five minutes for the larva to grow.

## Changelog
:cl:
code: Xenomorph embryos now take a static five minutes (1 minute per 5 stages) to hatch, instead of taking any time between fifteen seconds, till the end of time itself to hatch. Aka it had a 3% chance and rolled every two seconds, now it uses a consistent timer.
/:cl:

## Happy Boxing Day!
I tested this in-game and it's fully functional. Var-editing the variable the timer looks for to decide how long it should run works, I dunno why it wouldn't. Everything works!

**Dear Maintainers,**
If you don't want it to take 5 minutes I have absolutely no attachment to how long it actually takes, if you want it to take (insert time here) I'll happily change it, no arguments. I'm also perfectly fine with you just doing a webedit to the PR and changing the timer yourself to whatever you want.

**Dear TG, I guess.**
Hope you all had a wonderful Christmas!

KathrinBailey committed:
Makes alien embryos grow consistently instead of growing per utter random chance.

coiax commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -74,6 +87,7 @@
 	if(!candidates.len || !owner)
 		bursting = FALSE
 		stage = 4
+		advance_embryo_stage()	//We stop growing at stage 5, so if we go back a stage, we need to be able to advance again.
```
Won't this immediately poll the ghosts again? Since, the stage is 4, is immediately ++'d to 5.

Shouldn't this be instead a 60 timer to calling advance_embryo_stage?

KathrinBailey replied:
Good catch!
Thank you, it would do this. The proc was arranged differently when I made this change, and I evidently forgot to change this to reflect the proc's changes.

KathrinBailey committed:
Fixes bug

Makes it so going down to stage 4 if there's no candidates will start a timer to call the advance embryo stage proc instead of going directly to the proc, as it'd just go from 4, then the proc will kick it to 5 again immediately.

coiax commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -46,11 +54,16 @@
+	if(stage < 5)
 		stage++
 		INVOKE_ASYNC(src, .proc/RefreshInfectionImage)
+		addtimer(CALLBACK(src, .proc/advance_embryo_stage), growth_time)
```
Should this timer only be scheduled if the NEW stage is less than 5?

KathrinBailey replied:
Oh yeah. Nothing bad happens if you go above stage 5, but fixed anyway (I think).

coiax commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -46,11 +54,16 @@
 		stage++
 		INVOKE_ASYNC(src, .proc/RefreshInfectionImage)
+		addtimer(CALLBACK(src, .proc/advance_embryo_stage), growth_time)
+	return
```
Don't need this return at the end of the function, since this is the end of the function anyway.

KathrinBailey committed:
Update alien_embryo.dm

KathrinBailey committed:
Operator Derp

Rohesie commented:
This is a good PR, but you might want to talk with the author of #55937 to see if the plans match.

IndieanaJones commented:
I think this is just fine, I actually avoided messing with embryos because I saw this PR was up, but I had the same plans.

Rohesie commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -46,11 +54,16 @@
+/obj/item/organ/body_egg/alien_embryo/proc/advance_embryo_stage()
+	if(stage >= 5)
+		return
+	if(stage < 5)
```
No need for this check, given the prior one.

Rohesie commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -46,11 +54,16 @@
+	if(stage < 5)
 		stage++
 		INVOKE_ASYNC(src, .proc/RefreshInfectionImage)
+		if(stage >= 5)
+			addtimer(CALLBACK(src, .proc/advance_embryo_stage), growth_time)
```
I assume this check was forgotten here?

Rohesie commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -5,8 +5,16 @@
 	food_reagents = list(/datum/reagent/consumable/nutriment = 5, /datum/reagent/toxin/acid = 10)
-	var/stage = 0
+	//Which growth stage are we currently on? Stages give hints to the organ holder that this is in them.
+	var/stage = -1	//Once we initialise, we call our first growth cycle which immediately puts us to stage 0.
```
Either make it start at zero and increase the max stage value by one, or make this a define.

KathrinBailey committed:
Should be okay

KathrinBailey committed:
heck

KathrinBailey committed:
Update alien_embryo.dm

KathrinBailey committed:
big rohesie helped me do the thing cooler

KathrinBailey committed:
Update alien_embryo.dm

KathrinBailey committed:
Merge remote-tracking branch 'upstream/master' into blankets-for-little-xenos

KathrinBailey committed:
updating the now outdated documentation

KathrinBailey committed:
Update alien_embryo.dm

KathrinBailey committed:
updates documentation
